rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,857 @@
+  * Neither Chrome nor Firefox support EdDSA signatures
+  * The [CA Baseline Requirements](https://cabforum.org/baseline-requirements-documents/)
+    do not allow EdDSA signatures
+  * LibreSSL only added Ed25519 support in v3.7.0 (2022-12-12), the default
+    version of `curl` on my fully updated Macbook (Ventura 13.4.1) uses
+    LibreSSL 3.3.6 (2022-03-15)
+* YubiHSM and GCP KMS do *not* support Ed25519 keys.
+* YubiKey 5 does *not* support Ed25519 keys.
+* Ed25519 is considered by some to be the fastest, most secure, most modern
+  option for SSH certs.
+* Ed25519 key generation is *much* faster than RSA key generation.
+* Ed25519 signatures are faster than RSA signatures.
+* FIPS 186-5 approves Ed25519
+* Go BoringCrypto does not support Ed25519
+  <https://github.com/golang/go/blob/cd6676126b7e663e6202e98e2f235fff20d5e858/src/crypto/tls/boring.go#L78-L90>
+* Ed25519 is the only EdDSA curve supported in the Go standard library.
+
+#### Algorithms Summary
+
+* We are probably forced to continue unconditionally using RSA for database
+  certs, I'm assuming this would apply to both client and CA.
+* Ed25519 is a modern favourite for SSH, but TLS (and HSM, KMS) support is lacking.
+* Teleport CAs use separate keypairs for SSH and TLS, they do not need to use
+  the same algorithm.
+* Teleport derives client SSH and TLS certs from the same client keypair,
+  supporting different algorithms for each will require larger changes.
+* It seems it is time to split client SSH and TLS keys to support the popular
+  and secure Ed25519 algorithm for SSH and the widely-suported
+  `ECDSA_P256_SHA256` algorithm for TLS. This will also allows to evolve the
+  algorithms used for each protocol independently in the future.
+
+### UX
+
+The algorithm suite will be configurable in `teleport.yaml` and
+`cluster_auth_preference` as described above.
+
+We will add visibility of the current and configured CA algorithms to `tctl status`.
+
+```
+$ tctl status
+Cluster      cluster-one
+Version      14.0.0-dev
+
+Host CA pins:
+sha256:d2c825ede608da97891b3eaf1dac27e7e1085be5ff71e48a97ff8eb53360e362
+
+Host CA
+expires:         Jun 21 2033 00:03:19 UTC
+last rotated:    Jun 21 2023 00:03:19 UTC
+rotation state:  standby
+SSH algorithm:   Ed25519
+TLS algorithm:   ECDSA_P256_SHA256
+
+User CA
+expires:         Jun 21 2033 00:03:19 UTC
+last rotated:    never
+rotation state:  standby
+SSH algorithm:   RSA2048_PKCS1_SHA512 (balanced-v1 algorithm Ed25519 will take effect during next manual CA rotation)
+TLS algorithm:   RSA2048_PKCS1_SHA256 (balanced-v1 algorithm ECDSA_P256_SHA256 will take effect during next manual CA rotation)
+
+Database CA
+expires:         Jun 21 2033 00:03:19 UTC
+last rotated:    never
+rotation state:  rotating clients (mode: manual, started: Jun 30 2023 00:03:19 UTC, ending: Jul 1 2023 06:03:19 UTC)
+TLS algorithm:   RSA3072_PKCS1_SHA256
+
+...
+```
+
+`tctl auth rotate` will inform users when a rotation is going to trigger an
```

Comment:
What with a rollback? I assume that if rollback happens before we remove the old RSA key, this is not a problem. But what in a scenario when you rotate your key and after the rotation is completed, you realize that some of your existing tooling are not working, and you want to go back to RSA? Would that require switching to legacy and run the rotation again?

Reply:
added a mention of this in the doc